### PR TITLE
First stab on replacing links to other files in the outline with local links

### PR DESCRIPTION
**Pull request by DmitrySharabin**
The downside I noticed is that we replace all the links with the local ones, even those that are not supposed to be replaced (or should be ignored ?). Like this one:

SCR-20240819-jher

### CODE REVIEW

**Comment by LeaVerou** [comment]
An easy fix would be to introduce an outline-ignore class. Or, better, ignore links to the chapter itself since that’s clearly intentional.

**Comment by DmitrySharabin** [comment]
> Or, better, ignore links to the chapter itself since that’s clearly intentional.

It worked with the links to the Paged View inside separate chapters but not in the whole thesis, unfortunately. The code I wrote can’t distinguish, e.g., the “chapters/madata” href inside the Thesis Overview section and the one in the metadata under the Madata chapter heading in the whole thesis. So, they are treated the same—they are replaced with a local link.

@LeaVerou's answer from the Messenger:

Yeah I worried that would be the case. I think transforms operate on the whole page, so they don't have access to where each fragment came from. Perhaps the `outline-ignore` solution is the best here, at least until we think of a better solution (and a general escape hatch like `outline-ignore` is probably a good idea anyway).

File: src/Outlines.js [hunk 1]
```diff
@@ -412,8 +412,40 @@ export default class Outlines {
 			return open + info.label + " " + info.qualifiedNumber + groups.close;
 		});
 
-		// TODO replace links to other files in the outline with local links
+		// Replace links to other files in the outline with local links
 		// (whether they are empty or not)
+		content = content.replaceAll(hrefRegex, (match, ...args) => {
```
Comment (LeaVerou) [inline src/Outlines.js]: I think we have a `urls` field in one of the objects, right? Can we perhaps gate this behind a conditional so it's only executed when the outline actually spans multiple URLs?

Comment (DmitrySharabin) [reply src/Outlines.js]: I was using a condition with `urls` a bit later, but yeah, it makes more sense to do it earlier. Thanks! Fixed.

**Comment by LeaVerou** [comment]
Lgtm but given the amount of nav this breaks, can we implement the outline-ignore escape hatch in the same PR?

**Comment by DmitrySharabin** [comment]
> can we implement the outline-ignore escape hatch in the same PR

We definitely can. 🙂 Done! Could you please give it another look?

File: src/Outlines.js [hunk 2]
```diff
@@ -426,8 +427,48 @@ export default class Outlines {
...
+			// (whether they are empty or not)
+			content = content.replaceAll(linkRegex, (match, ...args) => {
+				let groups = processGroups(args.at(-1));
+				let attrs = groups.attrs;
+				attrs = html.parseAttributes(attrs);
+
+				let href = attrs.href;
+				let classes = attrs.class ?? "";
+
+				let ignore = href.startsWith("#") || classes.includes("outline-ignore");
```
Comment (DmitrySharabin) [inline src/Outlines.js]: I wonder if it makes sense to do something like this instead? 🤔

```js
let classes = (attrs.class ?? "").split(/[\s(\n\r?)]+/).filter(Boolean);
classes = new Set(classes);

let ignore = href.startsWith("#") || classes.has("outline-ignore");
```

Comment (LeaVerou) [reply src/Outlines.js]: or just `/(?:^|\s)outline-ignore(?:$|\s)/.test(classes)`

Comment (DmitrySharabin) [reply src/Outlines.js]: Love it! Will do it this way. Thanks!

Comment (DmitrySharabin) [reply src/Outlines.js]: Done!

**Review by LeaVerou — APPROVED** [review]

**Pull request merged**